fsck/fsck.c

```diff
@@ -668,6 +668,144 @@ static int check_name_dentry_set(struct exfat_de_iter *iter,
 	return 0;
 }
 
+static int check_bad_char(char w)
+{
+	return (w < 0x0020) || (w == '*') || (w == '?') || (w == '<') ||
+		(w == '>') || (w == '|') || (w == '"') || (w == ':') ||
+		(w == '/') || (w == '\\');
+}
+
+static char *get_rename_from_user(struct exfat_de_iter *iter)
+{
+	char *rename = malloc(ENTRY_NAME_MAX + 2);
+
+	if (!rename)
+		return NULL;
+
+retry:
+	/* +2 means LF(Line Feed) and NULL terminatior */
+	memset(rename, 0x1, ENTRY_NAME_MAX + 2);
+	printf("New name: ");
+	if (fgets(rename, ENTRY_NAME_MAX + 2, stdin)) {
+		int i, len, err;
+		struct exfat_lookup_filter filter;
+
+		len = strlen(rename);
+		/* Remove LF in filename */
+		rename[len - 1] = '\0';
+		for (i = 0; i < len - 1; i++) {
+			if (check_bad_char(rename[i])) {
+				printf("filename contain invalid character(%c)\n", rename[i]);
+				goto retry;
+			}
+		}
+
+		exfat_de_iter_flush(iter);
+		err = exfat_lookup_file(iter->exfat, iter->parent, rename, &filter);
+		if (!err) {
+			printf("file(%s) already exists, retry to insert name\n", rename);
+			goto retry;
+		}
+	}
+
+	return rename;
+}
+
+static char *generate_rename(struct exfat_de_iter *iter)
+{
+	char *rename;
+
+	if (iter->dot_name_num > DOT_NAME_NUM_MAX)
+		return NULL;
+
+	rename = malloc(ENTRY_NAME_MAX + 1);
+	if (!rename)
+		return NULL;
+
+	while (1) {
+		struct exfat_lookup_filter filter;
+		int err;
+
+		snprintf(rename, ENTRY_NAME_MAX + 1, "FILE%07d.CHK",
+			 iter->dot_name_num++);
+		err = exfat_lookup_file(iter->exfat, iter->parent, rename,
+					&filter);
+		if (!err)
+			continue;
+		break;
+	}
+
+	return rename;
+}
+
+const __le16 MSDOS_DOT[ENTRY_NAME_MAX] = {cpu_to_le16(0x002e), 0, };
+const __le16 MSDOS_DOTDOT[ENTRY_NAME_MAX] = {cpu_to_le16(0x002e), cpu_to_le16(0x002e), 0, };
+
+static int handle_dot_dotdot_filename(struct exfat_de_iter *iter,
+				      struct exfat_dentry *dentry)
+{
+	char *filename;
+	char error_msg[150];
+	int name_len, ret, num;
+
+	if (!memcmp(dentry->name_unicode, MSDOS_DOT, ENTRY_NAME_MAX * 2))
```

Comment:
If adding check that stream_name_len is 1 or 2, we can filter out dentries quickly.

Reply:
Done.

Reply:
Hm.. memcmp len is ENTRY_NAME_MAX * 2. I'm worried that this may cause a performance degrade. how about we only check 1 or 2 len as memcmp length ?

Reply:
Done. Please check the patch.

Reply:
Looks good to me.